examples/graph/gnn_citations.py

```diff
@@ -0,0 +1,728 @@
+        # neighbour_repesentations shape is [num_edges, representation_dim].
+        neighbour_repesentations = tf.gather(node_repesentations, neighbour_indices)
+
+        # Prepare the messages of the neighbours.
+        neighbour_messages = self.prepare(neighbour_repesentations, edge_weights)
+        # Aggregate the neighbour messages.
+        aggregated_messages = self.aggregate(node_indicies, neighbour_messages)
+        # Update the node embedding with the neighbour messages.
+        node_embeddings = self.update(node_repesentations, aggregated_messages)
+
+        return node_embeddings
+
+
+"""
+### Implement a graph neural network node classifier
+
+The GNN classification model follows the [Design Space for Graph Neural Networks](https://arxiv.org/abs/2011.08843) approach,
+as follows:
+
+1. Apply preprocessing using FFN to the node features to generate initial node representations.
+2. Apply one or more graph convolutional layer, with skip connections,  to the node representation
+to produce node embeddings.
+3. Apply post-processing using FFN to the node embeddings to generat the final node embeddings.
+4. Feed the node embeddings in a Softmax layer to predict the node class.
+
+Each graph convolutional layer added captures information from a further level of neighbours.
+However, adding many graph convolutional layer can cause oversmoothing, where the model
+produces similar embeddings for all the nodes.
+
+Note that the `graph_info` passed to the constructor of the Keras model, and used as a *property*
+of the Keras model object, rather than input data for training or prediction.
+The model will accept a **batch** of `node_indices`, which are used to lookup the
+node features and neighbours from the `graph_info`.
+"""
+
+
+class GNNNodeClassifier(tf.keras.Model):
+    def __init__(
+        self,
+        graph_info,
+        num_classes,
+        hidden_units,
+        aggregation_type="sum",
+        combination_type="concat",
+        dropout_rate=0.2,
+        normalize=True,
+        *args,
+        **kwargs,
+    ):
+        super(GNNNodeClassifier, self).__init__(*args, **kwargs)
+
+        # Unpack graph_info to three elements: node_features, edges, and edge_weight.
+        node_features, edges, edge_weights = graph_info
+
+        self.node_features = node_features
+        self.edges = edges
+        self.edge_weights = edge_weights
+        # Set edge_weights to ones if not provided.
+        if self.edge_weights is None:
+            self.edge_weights = tf.ones(shape=edges.shape[1])
+        # Scale edge_weights to sum to 1.
+        self.edge_weights = self.edge_weights / tf.math.reduce_sum(self.edge_weights)
+
+        self.num_classes = num_classes
+        self.hidden_units = hidden_units
+        self.aggregation_type = aggregation_type
+        self.combination_type = combination_type
+        self.dropout_rate = dropout_rate
+        self.normalize = normalize
+
+    def build(self, input_shape):
```

Comment:
Given that you don't use `input_shape`, prefer moving this to `__init__`. The general expectation is that calling `build` should create the weights of the network, which isn't the case here.

Reply:
Done